examples/vision/semisupervised_simclr.py

```diff
@@ -0,0 +1,743 @@
+    ]
+    plt.figure(figsize=(num_images * 2.2, 4 * 2.2), dpi=100)
+    for column, image_row in enumerate(augmented_images):
+        for row, image in enumerate(image_row):
+            plt.subplot(4, num_images, row * num_images + column + 1)
+            plt.imshow(image)
+            if column == 0:
+                plt.title(row_titles[row], loc="left")
+            plt.axis("off")
+    plt.tight_layout()
+
+
+visualize_augmentations(num_images=8)
+
+"""
+## Encoder architecture
+"""
+
+# define the encoder architecture
+def get_encoder(width):
+    return keras.Sequential(
+        [
+            layers.Input(shape=(96, 96, 3)),
+            layers.Conv2D(width, kernel_size=3, strides=2, activation="relu"),
+            layers.Conv2D(width, kernel_size=3, strides=2, activation="relu"),
+            layers.Conv2D(width, kernel_size=3, strides=2, activation="relu"),
+            layers.Conv2D(width, kernel_size=3, strides=2, activation="relu"),
+            layers.Flatten(),
+            layers.Dense(width, activation="relu"),
+        ],
+        name="encoder",
+    )
+
+
+"""
+## Supervised baseline model
+
+A baseline supervised model is trained using random initialization for 60 epochs.
+"""
+
+# baseline supervised training with random initialization
+baseline_model = keras.Sequential(
+    [
+        layers.Input(shape=(96, 96, 3)),
+        get_classification_augmenter(),
+        get_encoder(width),
+        layers.Dense(10),
+    ],
+    name="not_pretrained_model",
+)
+baseline_model.compile(
+    optimizer=keras.optimizers.Adam(),
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    metrics=[keras.metrics.SparseCategoricalAccuracy(name="acc")],
+)
+
+baseline_history = baseline_model.fit(
+    labeled_train_dataset, epochs=num_epochs, validation_data=test_dataset
+)
+print(
+    "Maximal validation accuracy: {:.2f}%".format(
+        max(baseline_history.history["val_acc"]) * 100
+    )
+)
+
+"""
+## Self-supervised model for contrastive pretraining
+
+We pretrain an encoder on unlabeled images with a contrastive loss for 30 epochs. A
+nonlinear projection head is attached to the top of the encoder, as it improves the
+quality of representations of the encoder.
+
+The InfoNCE/NT-Xent/N-pairs loss is used, which can be interpreted in the following way:
+- we treat each image in the batch as if it had its own class
+- then we have two examples (a pair of augmented views) for each "class"
+- each view's representation is compared to every possible pair's one (for both augmented
+versions)
+- we use the temperature-scaled cosine similarity of compared representations as logits
+- and use categorical cross-entropy as the "classification" loss
+
+The following two metrics are used for monitoring the pretraining performance:
+- **contrastive accuracy**: self-supervised metric, the ratio of cases in which the
+closest representation to a view was its differently augmented version from the same
```

Comment:
was different? Unsure what the sentence means.

Reply:
Yeah... I reworded it to the following: "the ratio of cases in which the representation of an image is more similar to its differently augmented version's one, than to the representation of any other image in the current batch".